src/libbson/src/bson/bson-error.c

```diff
@@ -98,25 +103,47 @@ bson_set_error (bson_error_t *error, /* OUT */
  */
 
 char *
-bson_strerror_r (int err_code,  /* IN */
-                 char *buf,     /* IN */
-                 size_t buflen) /* IN */
+bson_strerror_r (int err_code,                    /* IN */
+                 char *buf BSON_MAYBE_UNUSED,     /* IN */
+                 size_t buflen BSON_MAYBE_UNUSED) /* IN */
 {
    static const char *unknown_msg = "Unknown error";
    char *ret = NULL;
 
 #if defined(_WIN32)
+   // Windows does not provide `strerror_l` or `strerror_r`, but it does
+   // unconditionally provide `strerror_s`.
    if (strerror_s (buf, buflen, err_code) != 0) {
       ret = buf;
    }
-#elif defined(BSON_HAVE_XSI_STRERROR_R)
-   if (strerror_r (err_code, buf, buflen) == 0) {
-      ret = buf;
+#elif defined(__APPLE__)
+   // Apple does not provide `strerror_l`, but it does unconditionally provide
+   // the XSI-compliant `strerror_r`, but only when compiling with Apple Clang.
+   // GNU extensions may still be a problem if we are being compiled with GCC on
+   // Apple. Avoid the compatibility headaches with GNU extensions and the musl
+   // library by assuming QoI will not cause UB when reading the error message
+   // string even when `strerror_r` fails.
+   (void) strerror_r (err_code, buf, buflen);
```

Comment:
Should we consider using this code path for the final `#else` case instead of raising a build error of "Unable to find a supported strerror_r candidate"?

I'll note that PHP core does that (see: https://github.com/php/php-src/pull/11624#discussion_r1272237069). IMO it's not the ideal approach (I disagree with PHP doing it for _all_ cases, since they never bother to detect the XSI variant), but it generally works. And given that errno translation probably isn't a critical component of libbson, I think a user would be more inclined to receive "Unknown error" at runtime than a build error.

Looking at uses of `bson_strerror_r()` throughout libbson and libmongoc, I see that the `err_code` param is only _sometimes_ included in the logged error message. That seems like something that could be addressed independently. Rather than hunt through all of those occurrences, a compromise might be to incorporate it into the returned string for `!ret` at the end of this function (e.g. `"Unknown error (errno: %d)"`).

Reply:
> For example, bson_json_reader_new_from_file's errmsg_buf buffer is [uninitialized](https://github.com/mongodb/mongo-c-driver/blob/14edfd2c605be54e11ec133e04ca0e320dec3ec8/src/libbson/src/bson/bson-reader.c#L784). If strerror_r fails internally and does not write a failure message to the buffer, the successful result of bson_strerror_r would point to the uninitialized buffer as the error message string, and any subsequent reads of that error message string would be undefined behavior.

Ah, very good point. I was confused by jumping back and forth between libbson and PHP, which allocates the buffer in https://github.com/php/php-src/blob/3d832da8f53d2c2f93d4ffbd1c807d9a5d66536b/Zend/zend.c#L1631-L1641 (where the worst case scenario there is just an empty string).

SGTM.

Reply:
> where the worst case scenario there is just an empty string

FYI the linked function is stil susceptible to the same potential undefined behavior as described for `bson_json_reader_new_from_file`: the `buf` array is allocated but its elements are left uninitialized. Subsequent read of the `buf` elements in `zend_error_noreturn` may be undefined behavior if `strerror_r` failed and did not write a message to `buf`. The risk could be reduced by initializing the elements of the `buf` array, but it would still be a risk nevertheless (reliance on QoI to always write something to `buf` even on failure).

Reply:
Ah, right you are again. The `buf` is declared within the function and unallocated. I was thinking of a static/global decl being allocated with zeroes.